ruby/neo4j/driver/internal/util/error_util.rb

```diff
@@ -0,0 +1,113 @@
+                          when 'ClientError'
+                            if extract_error_sub_class(code) == 'Security'
+                              SEC_EXCEPTION_CODE_MAPPING[code.downcase] || Neo4j::Driver::Exceptions::SecurityException
+                            else
+                              code.casecmp?('Neo.ClientError.Database.DatabaseNotFound') ? Neo4j::Driver::Exceptions::FatalDiscoveryException : Neo4j::Driver::Exceptions::ClientException
+                            end
+                          when 'TransientError'
+                            Neo4j::Driver::Exceptions::TransientException
+                          else
+                            Neo4j::Driver::Exceptions::DatabaseException
+                          end
+
+        exception_class.new(code, message)
+      end
+
+      def fatal?(error)
+        return true unless error.is_a?(Neo4j::Driver::Exceptions::Neo4jException)
+
+        error_code = error.code&.downcase
+        return true if protocol_violation_error?(error_code)
+        return false if client_or_transient_error?(error_code)
+        # TO DO: last two if condition makes this method return nil, should we change it to only return boolean
+      end
+
+      def rethrow_async_exception(exception)
+        error = exception.cause
+        internal_cause = InternalExceptionCause.new(nil, error.backtrace)
+        error.add_suppressed(internal_cause) # TO DO: Ruby exception don't support supressed, how to handle this?
+
+        # do not include Thread.current and this method in the stacktrace
+        current_stack_trace = Thread.current.backtrace.drop(2)
+        error.set_backtrace(current_stack_trace)
+
+        org.neo4j.driver.internal.shaded.io.netty.util.internal.PlatformDependent.throw_exception(error)
+      end
+
+      def add_suppressed(main_error, error)
+        main_error.add_suppressed(error) if main_error != error
+      end
+
+      def get_root_cause(error)
+        java.util.Objects.require_non_null(error)
+
+        error.cause ? get_root_cause(error.cause) : error
+      end
+
+      # Exception which is merely a holder of an async stacktrace, which is not the primary stacktrace users are interested in.
+      # Used for blocking API calls that block on async API calls.
+      class InternalExceptionCause < RuntimeError
+        def initialize(message, backtrace)
+          super(message)
+          set_backtrace(backtrace)
+        end
+      end
+
+      private
+
+      def extract_error_class(code)
+        extract_class_from_code(code, 2)
+      end
+
+      def extract_error_sub_class(code)
+        extract_class_from_code(code, 3)
+      end
+
+      def extract_class_from_code(code, parts_counts)
+        parts = code.split('.')
+        parts.length < parts_counts ? '' : parts[parts_counts - 1]
+      end
+
+      def protocol_violation_error?(error_code)
+        return false if error_code.nil?
+
+        error_code.start_with?('neo.clienterror.request')
```

Comment:
That's overinterpreting. The error code comes from the server and is well defined.